TASK: Split basic fusion prototype files and define defaults

Sebobo opened the pull request:
Splits the Fusion prototypes in Neos.Neos and Neos.Fusion into separate files and adds all available parameters of the PHP implementations also in Fusion.
This should make it easier for integrators to understand which options are available for each prototype without reading the their PHP implementation code.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [x] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

Sebobo committed:
TASK: Split Fusion prototype files

This makes it easier to navigate to specific prototypes.
Deprecated prototypes are marked and now in a subfolder

Sebobo committed:
TASK: Add defaults in Fusion for all basic prototypes

This should make it more obvious which fusion parameters are
available for each prototype and inspecting the PHP code should
not be as necessary anymore.

mhsdesign commented:
NOICE! 🎉

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/Root.fusion:
```diff
@@ -1,2 +1,3 @@
-include: RawContentMode.fusion
\ No newline at end of file
+include: RootCase.fusion
+include: RawContentMode.fusion
```
how about we add `include: Prototypes/*.fusion` here

and remove it from RootCase :)

Sebobo committed:
TASK: Move Neos prototypes include to Root.fusion

Benjamin-K commented:
Very nice. 
I'd prefer having an example for each of the prototypes, especially for beginners. A beginner could guess what `Neos.Fusion:CanRender` or `Neos.Fusion:Component` does, but an example would help a lot, especially if we have it for some of the prototypes.

mhsdesign commented:
+1

but we should be carefull about the single source of truth - maybe generate the docs from the sourcecode?
https://discuss.neos.io/t/documenting-fusion/6056/4?u=marc

Sebobo commented:
I thought about the examples too, but would add them in follow-up PRs, as they are more work and better done in batches maybe.

mhsdesign commented:
i agree that can be done later. This IS already a big improvement to have all paths of the object written out ;)

mhsdesign commented:
i dont think we should start inline documentation, when this code would not be used for the rendered docs (as thoses two things can easily diverge).

but currently the fusion ast parser ignores any information about the comments - that would need to be changed...

Benjamin-K commented:
Then skip for now. We can add a link to the docs later instead of creating full examples inside the fusion files. 
This already is a big improvement 👍

mhsdesign reviewed (commented):
Seems to work ;)

Question:
Is it now api to include only a specific fusion file from Neos.Fusion?

like
`include: resources://Neos.Fusion/Private/Fusion/Value.fusion`

i would be against promoting that pattern - as it would not allow us to use the pattern of 
moving deprecated objects to `Deprecated` in the future.

but how can we tell the dev to  not do that but use `include: resources://Neos.Fusion/Private/Fusion/Root.fusion` ?

(same goes for `Neos.Neos` `Root.fusion` should be "api" everything else not.)

Benjamin-K commented on Neos.Fusion/Resources/Private/Fusion/Deprecated/Attributes.fusion:
```diff
@@ -0,0 +1,15 @@
+#   }
+# }
+#
+# @deprecated since Neos 5.0 in favor of Neos.Fusion:Tag or Neos.Fusion:Join
```
Shouldn't this be `in favor of Neos.Fusion:DataStructure`? Optionally with `@sortProperties = false`?

mhsdesign replied:
i guess - actually there is no real 1to1 replacement

i would suggest:
`in favor of Neos.Fusion:Tag internal attributes handling when a datastructure is passed`

Sebobo replied:
I took the deprecation notice from the PHP code. But I also prefer your version.

Benjamin-K commented on Neos.Fusion/Resources/Private/Fusion/Root.fusion:
```diff
@@ -1,291 +1 @@
-  format = ${request.format}
-  baseUri = ${String.toString(BaseUri.getConfiguredBaseUriOrFallbackToCurrentRequest(request.httpRequest))}
-}
+include: **/*.fusion
```
As Marc mentioned, I think we should move all Fusion files from root to a Prototypes folder. Then we can include only the folders we want and maybe remove `Deprecated` from the include.

Sebobo replied:
Done

Sebobo committed:
TASK: Adjust deprecation notice for Neos.Fusion:Attributes

Sebobo committed:
TASK: Move Neos.Fusion prototypes to subfolder

This allows including only the non deprecated prototypes
from other packages if that is desired.

Benjamin-K reviewed (approved):
Looks good now. 
Is there a plan for the removal of the auto-include of the `Deprecated` folder in `Neos.Fusion`?

mhsdesign commented:
yes we often had that plan ^^ it was even proposed to remove them directly next release (that would have been neos 5 ^^)
https://github.com/neos/neos-development-collection/pull/2189

so i guess its gonna be 9.0 since we missed it with 8.0

Benjamin-K commented:
What about a **soft removal** in 8.3 by only removing the include – so `Root.fusion` in `Neos.Fusion` only includes `Prototypes/*.fusion`? 
That way, developers can include the Deprecated folder if they still use old prototypes and in 9.0 we can remove them completely? So developers will have kind of a pre-warning that they need to adjust their prototypes.

mhsdesign commented:
even if this would be technically breaking - one can easily fix that as you said.
i like this better that completely destoying everything with 9.0 (ESCR is already quite breaky ^^)

... but on the other hand if we could provide a good migration (i think we have already one?) then it should be quite easy to adjust the code and should not require a special soft removal?

Sebobo commented:
There is a migration and we should target 9.0 for this.
I don't see how those few prototypes could be a problem until then.

mhsdesign commented:
we missed `renderer` in https://github.com/neos/neos-development-collection/blob/9c7daf8790096f41496ba244a519f302bddcd388/Neos.Fusion/Resources/Private/Fusion/Prototypes/Matcher.fusion#L5

Sebobo commented:
Shame shame shame!

mhsdesign commented:
😂😂
noticed it randomly when using your declarations to teach the api of the matcher

too bad i guess fusion will collapse now and burst into fire. Ill fix it once im on my pc ^^

mhsdesign commented on Neos.Fusion/Resources/Private/Fusion/Prototypes/Component.fusion:
```diff
@@ -0,0 +1,5 @@
+prototype(Neos.Fusion:Component) {
+  @class = 'Neos\\Fusion\\FusionObjects\\ComponentImplementation'
+  @sortProperties = false
+  renderer = ''
```
we need to change that to `renderer = null`, as otherwise the path is existant and one doesnt get an error if its missed.

mhsdesign replied:
will be fixed via https://github.com/neos/neos-development-collection/pull/3945

mhsdesign commented:
Ups we may have broke the default error handling in 8.2 https://discuss.neos.io/t/redirect-issue-to-404-not-found-page-in-neos-demo-neos-flow-8-2-0/6169/3?u=marc

… needs investigation ;)
